sympy/solvers/ode/single.py

```diff
@@ -359,6 +358,150 @@ def inverse(self):
+    differential of a function. That is, the differential equation
+
+    .. math:: P(x, y) \,\partial{}x + Q(x, y) \,\partial{}y = 0
+
+    is exact if there is some function `F(x, y)` such that `P(x, y) =
+    \partial{}F/\partial{}x` and `Q(x, y) = \partial{}F/\partial{}y`.  It can
+    be shown that a necessary and sufficient condition for a first order ODE
+    to be exact is that `\partial{}P/\partial{}y = \partial{}Q/\partial{}x`.
+    Then, the solution will be as given below::
+
+        >>> from sympy import Function, Eq, Integral, symbols, pprint
+        >>> x, y, t, x0, y0, C1= symbols('x,y,t,x0,y0,C1')
+        >>> P, Q, F= map(Function, ['P', 'Q', 'F'])
+        >>> pprint(Eq(Eq(F(x, y), Integral(P(t, y), (t, x0, x)) +
+        ... Integral(Q(x0, t), (t, y0, y))), C1))
+                    x                y
+                    /                /
+                   |                |
+        F(x, y) =  |  P(t, y) dt +  |  Q(x0, t) dt = C1
+                   |                |
+                  /                /
+                  x0               y0
+
+    Where the first partials of `P` and `Q` exist and are continuous in a
+    simply connected region.
+
+    A note: SymPy currently has no way to represent inert substitution on an
+    expression, so the hint ``1st_exact_Integral`` will return an integral
+    with `dy`.  This is supposed to represent the function that you are
+    solving for.
+
+    Examples
+    ========
+    >>> from sympy import Symbol,Function
+    >>> from sympy import E,dsolve,Eq,log,sin,cos
+    >>> x=Symbol('x')
+    >>> f=Function('f')
+    >>> eq=Eq(0,(sin(x)*sin(f(x))-x*E**f(x))*f(x).diff(x)-(E**f(x)+cos(x)*cos(f(x))))
+    >>> dsolve(eq,hint='1st_exact')
+    Eq(x*exp(f(x)) + sin(x)*cos(f(x)), C1)
+    >>> dsolve(cos(f(x)) - (x*sin(f(x)) - f(x)**2)*f(x).diff(x),f(x), hint='1st_exact')
+    Eq(x*cos(f(x)) + f(x)**3/3, C1)
+
+    References
+    ==========
+
+    - "Differential Equations With Applications And Historical Notes",
+       Second Edition, George.F.Simmons
+    - https://en.wikipedia.org/wiki/Exact_differential_equation
+
+    # indirect doctest
+
+    """
+    hint="1st_exact"
+    has_integral=True
+    y = Dummy('y') # This is the only way to pass dummy y to _handle_Integral
+
+    def _wilds(self, f, x, order):
+        P = Wild('P')
+        Q = Wild('Q', exclude=[f(x).diff(x)])
+        return P, Q
+
+    def _equation(self, fx, x, order):
+        P, Q = self.wilds()
+        return P + Q*fx.diff(x)
+
+    def _verify(self, fx) -> bool:
+        P, Q = self.wilds()
+        eq = self.ode_problem.eq
+        x = self.ode_problem.sym
+        y = self.y
+
+        #This checks if Derivative present in wild P
+        if fx.diff(x) in self._wilds_match[P].atoms(Derivative):
+            eq = collect(eq,fx.diff(x))
+            pattern = self._equation(fx, x, 1)
+            self._wilds_match = eq.match(pattern)
```

Comment:
I don't understand what this is doing but it doesn't seem right

Reply:
The `_wilds_match` variable should be set by `PatternSolver.match`.

Reply:
For the case where `Derivative` is present in more then one terms, `eq` needs to be preprocessed before matching. This condition is required to process and verify that `P` is the function of `x` , ` f(x)` only and independent of any derivative.

Reply:
I think that should be handled by excluding `f(x).diff(x)` in the pattern

Reply:
The reason for not excluding `f(x).diff(x)` in `Wild P` and the need for this condition here are the same. Lets take an example of exact equation:
```
>>> P1 = Wild('P1',exclude=[f(x).diff(x)])
>>> Q1 = Wild('Q1',exclude=[f(x).diff(x)])
>>> 
>>> P2 = Wild('P2')
>>> Q2 = Wild('Q2',exclude=[f(x).diff(x)])
>>> 
>>> exprn1= P1 + Q1*f(x).diff(x)
>>> exprn2= P2 + Q2*f(x).diff(x)
>>> 
>>> eq=x*f(x).diff(x)+(2/f(x))*f(x).diff(x)+f(x)
>>> eq = eq.collect(f(x), func = cancel)
>>> 
>>> print(eq.match(exprn1))
None
>>> eq.match(exprn2)
{Q2_: 2/f(x), P2_: x*Derivative(f(x), x) + f(x)}
```

Here, match with `exprn1` returns `None`. Hence, excluding `f(x).diff(x)` in `Wild P` will fail the `matches` function for for the case where `f(x).diff(x)` is present in more then one terms of `eq`.
After we match if `f(x).diff(x)` is present in `P`, we update it in above `if` block in  `_verify` function.

Reply:
Every PatternODESolver will have different preprocess for matching. We need a `_preprocess` function in `SinglePatternODESolver` base class which preprocesses equation before it is matched.

Reply:
We need to collect on the derivatives of f as well:
```julia
In [7]: eq.collect(f(x).diff(x))                                                                                                  
Out[7]: 
⎛     2  ⎞ d              
⎜x + ────⎟⋅──(f(x)) + f(x)
⎝    f(x)⎠ dx             

In [8]: eq.collect(f(x).diff(x)).match(exprn1)                                                                                    
Out[8]: 
⎧                   2  ⎫
⎨P₁: f(x), Q₁: x + ────⎬
⎩                  f(x)⎭
```

Reply:
Absolutely. We need to collect on the derivatives of f, but this is not required by other solvers. Therefore, putting `eq.collect(f(x).diff(x))` in `_matches` of `SinglePatternODESolver` will not be a wise option.
We need `_preprocess_eq` in `SinglePatternODESolver`:
```
#Subclass should define here all preprocess on the equation required before pattern matching.
def _preprocess_eq(self, eq):
   prep_eq=eq
    return prep_eq
```
`eq = _preprocess_eq(eq)` will be called before `eq.match(pattern)` here:
https://github.com/sympy/sympy/blob/233cafb34b853f5fa5567ae0aee53c714a464270/sympy/solvers/ode/single.py#L242

After `_preprocess_eq` is defined in `SinglePatternODESolver`, we would implement it in `FirstExact` as:
```
def _preprocess_eq(self, eq):
    prep_eq = eq.collect(f(x).diff(x))
    return prep_eq
```
This will solve the problem of specific preprocessing required by different solvers before pattern matching

Reply:
We need to minimise solver-specific logic. If collecting on `f(x).diff(x)` is generally useful then it should be done in the `SingleODEPatternSolver` class. Is there any reason *not* to put it there?

Normally collecting on `f(x)` also collects on derivatives e.g.:
```julia
In [60]: eq = f(x).diff(x) + x*f(x).diff(x) + f(x) + x*f(x)                                                                       

In [61]: eq                                                                                                                       
Out[61]: 
           d                 d       
x⋅f(x) + x⋅──(f(x)) + f(x) + ──(f(x))
           dx                dx      

In [62]: eq.collect(f(x))                                                                                                         
Out[62]: 
                       d       
(x + 1)⋅f(x) + (x + 1)⋅──(f(x))
                       dx 
```
The reason that doesn't work in your example is because you have mixed products involving both `f(x)` and `f(x).diff(x)`:
```julia
In [63]: eq = f(x).diff(x) + (x/f(x))*f(x).diff(x) + f(x) + x*f(x)                                                                

In [64]: eq                                                                                                                       
Out[64]: 
           d                         
         x⋅──(f(x))                  
           dx                d       
x⋅f(x) + ────────── + f(x) + ──(f(x))
            f(x)             dx      

In [65]: eq.collect(f(x))                                                                                                         
Out[65]: 
  d                                 
x⋅──(f(x))                          
  dx                        d       
────────── + (x + 1)⋅f(x) + ──(f(x))
   f(x)                     dx      

In [66]: eq.collect(f(x)).collect(f(x).diff(x))                                                                                   
Out[66]: 
               ⎛ x      ⎞ d       
(x + 1)⋅f(x) + ⎜──── + 1⎟⋅──(f(x))
               ⎝f(x)    ⎠ dx 
```
I think that the right way to do this is like this in the first place:
```julia
In [68]: eq.collect([f(x).diff(x), f(x)])                                                                                         
Out[68]: 
               ⎛ x      ⎞ d       
(x + 1)⋅f(x) + ⎜──── + 1⎟⋅──(f(x))
               ⎝f(x)    ⎠ dx 
```
For ODE pattern matching collecting on the derivative is more important than collecting on `f(x)`.

Reply:
`eq.collect([f(x).diff(x)], func = cancel)` solves the problem for preprocess for all solvers including a second order Liouville solver for which I will open a PR separately.